### ISSUE TITLE
GenericContainer#withCopyFileToContainer is not efficient memory-wise

### ISSUE THREAD

**Issue by vmassol** [testcontainers/testcontainers-java#4203]
Hi,

We've recently seen OOM in our tests and after debugging  by generating a heap dump and opening it, we found that for the following code that we use in XWiki tests, TC is copying all the content of the `sourceDirectory` in memory:

```
MountableFile mountableDirectory = MountableFile.forHostPath(sourceDirectory);
container.withCopyFileToContainer(mountableDirectory, targetDirectory);
```

Since our directory contains 300MB+ of files, the memory we give to the JVM is not enough.

Note that we copy and don't bind because we need to have it work in the DOOD use case. Our code is:

```
    protected void mountFromHostToContainer(GenericContainer<?> container, String sourceDirectory,
        String targetDirectory)
    {
        // Note 1: File mounting is awfully slow on Mac OSX. For example starting Tomcat with XWiki mounted takes
        // 45s+, while doing a COPY first and then starting Tomcat takes 8s (+5s for the copy).
        // Note 2: For the DOOD use case, we also do the copy instead of the volume mounting since that would require
        // to have the sourceDirectory path mounted from the host and this would put and leave files on the host which
        // would not work with parallel executions (think about multiple CI jobs executing in parallel on the same host)
        // and would also not be clean.
        String osName = System.getProperty("os.name").toLowerCase();
        if (isInAContainer() || osName.startsWith("mac os x")) {
            MountableFile mountableDirectory = MountableFile.forHostPath(sourceDirectory);
            container.withCopyFileToContainer(mountableDirectory, targetDirectory);
        } else {
            container.withFileSystemBind(sourceDirectory, targetDirectory);
        }
    }
```

It would be great if TC could stream the data instead of keeping it in memory. Is that possible?

Thanks

**Comment by REslim30** [comment]
I'm currently look into this issue.

Can reproduce the error. Getting OOM errors for files 100mb locally.

Just from what I've gathered, it seems like it's possible to stream to container, I think under the hood it uses:

https://docs.docker.com/engine/api/v1.41/#tag/Container/operation/PutContainerArchive

I think the problem is that we create the Tar archive in memory - and so the whole file gets loaded into memory. One possible solution is to stream the tar archive to disk and then stream it to docker? Still looking into solutions though.

**Comment by kiview** [comment]
@REslim30 Thanks for looking into this, that sounds indeed like the right direction. @eddumelendez might share some more details since he was recently looking into a similar topic in the context of Jib integration.

**Comment by vmassol** [comment]
Hello guys. FTR we're still getting this issue on the XWiki build.

```
01:36:19,047 [ERROR] There was an error in the forked process
01:36:19,047 [ERROR] Java heap space
01:36:19,047 [ERROR] java.lang.OutOfMemoryError: Java heap space
01:36:19,047 [ERROR]    at java.base/java.util.Arrays.copyOf(Arrays.java:3745)
01:36:19,047 [ERROR]    at java.base/java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:120)
01:36:19,047 [ERROR]    at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:95)
01:36:19,047 [ERROR]    at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:156)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.CountingOutputStream.write(CountingOutputStream.java:62)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream$BufferAtATimeOutputChannel.write(FixedLengthBlockOutputStream.java:91)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.writeBlock(FixedLengthBlockOutputStream.java:259)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.maybeFlush(FixedLengthBlockOutputStream.java:169)
01:36:19,047 [ERROR]    at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.write(FixedLengthBlockOutputStream.java:206)
01:36:19,047 [ERROR]    at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.write(TarArchiveOutputStream.java:713)
01:36:19,047 [ERROR]    at java.base/java.io.InputStream.transferTo(InputStream.java:705)
01:36:19,047 [ERROR]    at java.base/java.nio.file.Files.copy(Files.java:3119)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:362)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:371)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:371)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.recursiveTar(MountableFile.java:371)
01:36:19,047 [ERROR]    at org.testcontainers.utility.MountableFile.transferTo(MountableFile.java:327)
01:36:19,047 [ERROR]    at org.testcontainers.containers.ContainerState.copyFileToContainer(ContainerState.java:306)
01:36:19,047 [ERROR]    at org.testcontainers.containers.ContainerState.copyFileToContainer(ContainerState.java:282)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer$$Lambda$603/0x00000001003ca040.accept(Unknown Source)
01:36:19,047 [ERROR]    at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:430)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:344)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer$$Lambda$571/0x00000001003a1440.call(Unknown Source)
01:36:19,047 [ERROR]    at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:334)
01:36:19,047 [ERROR]    at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.DockerTestUtils.startContainerInternal(DockerTestUtils.java:218)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.DockerTestUtils.startContainer(DockerTestUtils.java:181)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.AbstractContainerExecutor.start(AbstractContainerExecutor.java:56)
01:36:19,047 [ERROR]    at org.xwiki.test.docker.internal.junit5.servletengine.ServletContainerExecutor.startContainer(ServletContainerExecutor.java:323)
01:36:19,048 [ERROR]    at org.xwiki.test.docker.internal.junit5.servletengine.ServletContainerExecutor.start(ServletContainerExecutor.java:155)
01:36:19,048 [ERROR]
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:628)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:285)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:250)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1203)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1055)
01:36:19,048 [ERROR]    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:871)
```

**Comment by jandry** [comment]
Yes I have issue on 1.19.0, 1.18.x, only working with 1.18.0

**Comment by jandry** [comment]
I proposed a working workaround here https://github.com/testcontainers/testcontainers-java/issues/7239#issuecomment-1746603941

**Comment by jandry** [comment]
Seems same as https://github.com/testcontainers/testcontainers-java/issues/7239

**Comment by jandry** [comment]
Seems same as https://github.com/testcontainers/testcontainers-java/issues/7239

**Comment by thackel** [comment]
I am now also hit by this when using 1.19.1

**Currently it is not possible to copy larger files or a directory into the container without getting an OOM.**

There are a couple of other tickets which are also related to this OOM (#1348, #7239,#2863, #4203), so I think this is not an edge usecase.

**Comment by wimdeblauwe** [comment]
I just tried to upgrade to 1.19.3 (from 1.17.6) and I am also hit with `OutOfMemoryError: Java heap space`:

```
org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
Java heap space
java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.Arrays.copyOf(Arrays.java:3541)
	at java.base/java.io.ByteArrayOutputStream.toByteArray(ByteArrayOutputStream.java:187)
	at org.testcontainers.containers.ContainerState.copyFileToContainer(ContainerState.java:337)
	at org.testcontainers.containers.ContainerState.copyFileToContainer(ContainerState.java:308)
	at org.testcontainers.containers.GenericContainer$$Lambda/0x00007f8cc0cd0860.accept(Unknown Source)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:986)
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:444)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:357)
	at org.testcontainers.containers.GenericContainer$$Lambda/0x00007f8cc0cc4000.call(Unknown Source)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:347)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:333)
	at org.testcontainers.containers.ContainerisedDockerCompose.invoke(ContainerisedDockerCompose.java:64)
	at org.testcontainers.containers.ComposeDelegate.runWithCompose(ComposeDelegate.java:254)
	at org.testcontainers.containers.ComposeDelegate.createServices(ComposeDelegate.java:163)
	at org.testcontainers.containers.DockerComposeContainer.start(DockerComposeContainer.java:137)
```

**Comment by gba-foundever** [comment]
A fix proposal has been pushed here https://github.com/testcontainers/testcontainers-java/issues/7239
it's working for me

**Comment by vmassol** [comment]
@gba-foundever Do you mean this patch: https://github.com/alexanderankin/testcontainers-java/commit/21cbc05d6633bcff443478c7844851fbf5e1ea66 ? Thx

**Comment by gba-foundever** [comment]
@vmassol yes

**Comment by vmassol** [comment]
@kiview Hello! Any chance of having the patch at https://github.com/alexanderankin/testcontainers-java/commit/21cbc05d6633bcff443478c7844851fbf5e1ea66 be included in the next release of TC by any chance? :) Thx

**Comment by vmassol** [comment]
Does someone have a TC fork with the patch applied and released? :)

**Comment by wimdeblauwe** [comment]
Is it possible to give some priority to this? I believe putting the docker compose file in the root of the project is something a lot of people do (Even start.spring.io generates it like that). All the duplicates of this issue (#7239, #1348, #2863) are a testament to this being quite an issue for a lot of people currently.

Even some mechanism to tell testcontainers to not copy every file, but just `.env` for example would already be great I guess.
